industrial_ci/src/workspace.sh

```diff
@@ -242,7 +242,7 @@ function ici_install_dependencies {
       rosdep_opts+=(--skip-keys "$skip_keys")
     fi
     set -o pipefail # fail if rosdep install fails
-    ROS_PACKAGE_PATH=$cmake_prefix_path ici_exec_in_workspace "$extend" "." rosdep install "${rosdep_opts[@]}" | { grep "executing command" || true; }
+    ROS_PACKAGE_PATH=$cmake_prefix_path:${ROS_PACKAGE_PATH} ici_exec_in_workspace "$extend" "." rosdep install "${rosdep_opts[@]}" | { grep "executing command" || true; }
```

Comment:
`${ROS_PACKAGE_PATH}` should be empty at this point

Reply:
In the case of the nightly image it is not empty https://github.com/osrf/docker_images/pull/328.
But it is a hack because of a rosdep issue, so I dont think this commit should be merged in industrial_ci, this is only propagating the workaround to have a working example for this discussion

Reply:
I see. Shouldn't this path be included in `CMAKE_PREFIX_PATH`? I stumbled upon this issue as well and `CMAKE_PREFIX_PATH` did the trick in my tests. However, I haven't tested it against the nightly build yet.

Reply:
hmm I think rosdep does only look in `ROS_PACKAGE_PATH` and `AMENT_PREFIX_PATH`.
If all the packages in the nightly archive were registered in the ament index it would work (ongoing discussion at https://github.com/ros2/tinyxml_vendor/pull/14)

Reply:
> hmm I think rosdep does only look in ROS_PACKAGE_PATH

That's why I set it to `CMAKE_PREFIX_PATH`, this works (or worked?) for catkin and ament :)

Reply:
But `rosdep` is not Catkin/Ament, correct?

I haven't checked, but if `rosdep` is used to resolve package dependencies in the scenarios @mikaelarguedas is trying to support, `CMAKE_PREFIX_PATH` may not help.

Reply:
> in the scenarios @mikaelarguedas is trying to support, CMAKE_PREFIX_PATH may not help

I think I already tested it against this scenario, but I will test again :)

Reply:
> But rosdep is not Catkin/Ament, correct?

rosdep is rosbuild-based with some work-arounds for catkin and ament ;)

Reply:
> That's why I set it to CMAKE_PREFIX_PATH, this works (or worked?) for catkin and ament :)

So I looked a bit into that.
TL;DR this line can stay as is if we remove the COLCON_IGNORE file located at the root of the archive https://github.com/mikaelarguedas/industrial_ci/commit/241216c2206749df874f776fd89eda94040a3296
I dont think this file serve much purpose in this case and if removed it should likely be removed upstream either in the docker image or in the archive generation itself

---

More technical dive:

By setting `ROS_PACKAGE_PATH` to anything, we are basically asking `rosdep` (and under the hood `catkin_pkg`) to crawl these directories looking for catkin packages.
By setting it to the `CMAKE_PREFIX_PATH` that is in general higher up in the hierarchy than `ROS_PACKAGE_PATH` we're adding overhead.

`ament` by opposition uses the ament resource index using the prefixes listed in only `AMENT_PREFIX_PATH` and does not crawl or populate third-party variables like `CMAKE_PREFIX_PATH`. So the ROS2 docker images (e.g. `ros:dashing`) do not set `CMAKE_PREFIX_PATH` or `ROS_PACKAGE_PATH`. Only `AMENT_PREFIX_PATH`.


There seem to be a couple different scenarios for ros2 builds:
  - For ROS2 docker images with only debs: `CMAKE_PREFIX_PATH` is not set
    - So if this works in this case is because the `AMENT_PREFIX_PATH` is used
  - For workspaces built with colcon: `CMAKE_PREFIX_PATH` is set to `AMENT_PREFIX_PATH` (e.g. `/opt/ros/foxy`) for compatibility, **BUT** a `COLCON_IGNORE` file is dropped at the same location. So `catkin_pkg` thus `rosdep` cannot find the packages at that location
  - The nightly image is the same as the previous case + `ROS_PACKAGE_PATH` is set to the share directory (like in ROS1) `/opt/ros/foxy/share` so there is no `COLCON_IGNORE` and it can find the packages by crawling

Ideal outcome IMO: for ROS2 the crawling approach is never used. No need to set or modify `ROS_PACKAGE_PATH` or `CMAKE_PREFIX_PATH`, only use `AMENT_PREFIX_PATH`
Packages in the archive should all be registered in the ament index by either creating a marker file in the resource index or by being ament packages to start with

Reply:
>  only use AMENT_PREFIX_PATH

catkin packages do not set `AMENT_PREFIX_PATH` (?)

> So the ROS2 docker images (e.g. ros:dashing) do not set CMAKE_PREFIX_PATH

I know, and they should not set it. I am reading it from `setup.bash`
This is a temporary hack until rosdep can crawl the ament index.

Reply:
> catkin packages do not set AMENT_PREFIX_PATH (?)

not that I know of. Even if a tool build tool like colcon did it, I wouldn't expect catkin packages to register themselves in the index. 

>  This is a temporary hack until rosdep can crawl the ament index.

This is now a thing https://github.com/ros-infrastructure/rosdep/pull/699
The nightly image sets the `ROS_PACKAGE_PATH` only because rosdep cannot find non-ament packages like https://github.com/ros2/tinyxml_vendor/pull/14

Reply:
> TL;DR this line can stay as is if we remove the COLCON_IGNORE

I just tested it, the `COLCON_IGNORE` is breaking the work-around.

If `rosdep` now support `AMENT_PREFIX_PATH`, then we don't need to set `ROS_PACKAGE_PATH`.

Reply:
#456 now reverts the work-around properly.

Reply:
> I just tested it, the COLCON_IGNORE is breaking the work-around.

In what scenario does it break? Can you clarify how it breaks?

If removing it has a negative effect, I'll reject https://github.com/osrf/docker_images/pull/345